snuba/clickhouse/sql.py

```diff
@@ -33,3 +33,15 @@ def format_sql(self, format: Optional[str] = None) -> str:
         if format is not None:
             query = f"{query} FORMAT {format}"
         return query
+
+    def _sql_data_impl(self) -> List[Tuple[str, str]]:
+        """
+        Returns a list of parsed SQL elements for logging, must be
+        implemented by the subclass to provide functionality to the ``sql_data``
+        function.
+        """
+        raise NotImplementedError
+
+    def sql_data(self) -> List[Tuple[str, str]]:
```

Comment:
Since this is being written to the span through the set_tag method (which fills in an dictionary), I don't think you are gaining much by returning a List of tuples to preserve the order.
Why not just returning a dataclass that defines all the relevant fields so this is method would be less coupled with the fact it is only used for that specific metric today and generally more useful.
dataclasses have an asdict method that would turn the structure into a dictionary you can use to fill in the span data iterating like you do now for the list of tuples.

Reply:
The ordering is important because the sql_data is being used by the SQL generator for the ASTQuery. I could go back to iterating over an ordered list of keys when generating the code, though. Or, keep the two separate as copy-pasta.

The ordering, though, is probably still useful if I combine it with your other suggestions.